File: Source/Game/CollisionDetection.cs
using System;
using WCSharp.Api;
using static WCSharp.Api.Common;

public static class CollisionDetection
{
    public const float DEFAULT_WOLF_COLLISION_RADIUS = 73.5f;
    private const float CIRCLE_COLLISION_RADIUS = 78.0f;

    private static Predicate<Relic> IsBeaconOfUnitedLifeforce = r => r is BeaconOfUnitedLifeforce;

    private static Func<bool> WolfCollisionFilter(Kitty k)
    {
        return () =>
        {
            if (!k.Alive) return false;
            var filterUnit = GetFilterUnit();
            return filterUnit.UnitType == Constants.UNIT_CUSTOM_DOG && filterUnit.Alive;
        };
    }

    private static Func<bool> ShadowRelicWolvesFilter(ShadowKitty sk)
    {
        return () =>
        {
            return (GetUnitTypeId(GetFilterUnit()) == Constants.UNIT_CUSTOM_DOG)
                    && sk.Unit.Alive;
        };
    }

    private static Func<bool> ShadowRelicCircleFilter(ShadowKitty sk)
    {
        return () =>
        {
            var filterUnit = GetFilterUnit();
            return GetUnitTypeId(filterUnit) == Constants.UNIT_KITTY_CIRCLE
                    && filterUnit.Owner != sk.Player // Not Same Player
                    && sk.Unit.Alive // Has to Be Alive
                    && Globals.ALL_KITTIES[filterUnit.Owner].TeamID == Globals.ALL_KITTIES[sk.Player].TeamID; // Must be same team

        };
    }

    private static Func<bool> CircleCollisionFilter(Kitty k)
    {

        return () =>
        {
            var kTeamID = k.TeamID;
            var kPlayer = k.Player;
            if (Gamemode.CurrentGameMode == GameMode.Solo) return false;
            if (!k.Alive) return false;
            var filterUnit = GetFilterUnit();
            if (filterUnit.UnitType != Constants.UNIT_KITTY_CIRCLE) return false;

            if (filterUnit.Owner == kPlayer) return false; // Not Same Player

            var filterTeamId = Globals.ALL_KITTIES[filterUnit.Owner].TeamID;
            return filterTeamId == kTeamID;
        };
    }

    public static void KittyRegisterCollisions(Kitty k)
    {
        var WOLF_COLL_RADIUS = k.CurrentStats.CollisonRadius;
        k.w_Collision ??= CreateTrigger();
        k.c_Collision ??= CreateTrigger();

        UnitWithinRange.RegisterUnitWithinRangeTrigger(k.Unit, WOLF_COLL_RADIUS, WolfCollisionFilter(k), WolfCollisionTrigger(k));
        UnitWithinRange.RegisterUnitWithinRangeTrigger(k.Unit, CIRCLE_COLLISION_RADIUS, CircleCollisionFilter(k), CircleCollisionTrigger(k));
    }

    public static void ShadowKittyRegisterCollision(ShadowKitty sk)
    {
        var WOLF_COLL_RADIUS = sk.Kitty.CurrentStats.CollisonRadius;
        sk.wCollision ??= CreateTrigger();
        sk.cCollision ??= CreateTrigger();

        UnitWithinRange.RegisterUnitWithinRangeTrigger(sk.Unit, WOLF_COLL_RADIUS, ShadowRelicWolvesFilter(sk), WolfCollisionShadowTrigger(sk));
        UnitWithinRange.RegisterUnitWithinRangeTrigger(sk.Unit, CIRCLE_COLLISION_RADIUS, ShadowRelicCircleFilter(sk), CircleCollisionShadowTrigger(sk));
    }

    private static trigger WolfCollisionTrigger(Kitty k)
    {
        TriggerAddAction(k.w_Collision, () =>
        {
            try
            {
                if (!k.Unit.Alive) return;
                if (k.Invulnerable) return; // proc before rewind and logic wise

                var filterUnit = GetFilterUnit();
                if (Globals.ALL_WOLVES[filterUnit].IsReviving) return; // bomber wolf
                if (NamedWolves.ExplodingWolfCollision(filterUnit, k)) return;
                if (ChronoSphere.RewindDeath(k)) return;
                OneOfNine.OneOfNineEffect(k);
                k.KillKitty();
                TeamsUtil.CheckTeamDead(k);
            }
            catch (Exception e)
            {
                Logger.Warning($"WolfCollisionTrigger Error: {e.Message}");
                throw;
            }
        });
        return k.w_Collision;
    }

    private static trigger CircleCollisionTrigger(Kitty k)
    {
        TriggerAddAction(k.c_Collision, () =>
        {
            try
            {
                var circle = Globals.ALL_KITTIES[GetFilterUnit().Owner];
                circle.ReviveKitty(k);
                (k.Relics.Find(IsBeaconOfUnitedLifeforce) as BeaconOfUnitedLifeforce)?.BeaconOfUnitedLifeforceEffect(k.Player);
            }
            catch (Exception e)
            {
                Logger.Warning($"CircleCollisionTrigger Error: {e.Message}");
                throw;
            }
        });
        return k.c_Collision;
    }

    private static trigger WolfCollisionShadowTrigger(ShadowKitty sk)
    {
        TriggerAddAction(sk.wCollision, () =>
        {
            if (NamedWolves.ExplodingWolfCollision(GetFilterUnit(), sk.Kitty, true)) return; // Floating text will appear on kitty instead of SK tho.
            sk.KillShadowKitty();
        });
        return sk.wCollision;
    }

    private static trigger CircleCollisionShadowTrigger(ShadowKitty sk)
    {
        TriggerAddAction(sk.cCollision, () =>
        {
            try
            {
                var circle = Globals.ALL_KITTIES[GetOwningPlayer(GetFilterUnit())];
                var saviorKitty = Globals.ALL_KITTIES[sk.Player];
                circle.ReviveKitty(saviorKitty);
            }
            catch (Exception e)
            {
                Logger.Warning($"CircleCollisionShadowTrigger Error: {e.Message}");
                throw;
            }
        });
        return sk.cCollision;
    }
}

File: src/Game/CollisionDetection.ts
export class CollisionDetection {
    public DEFAULT_WOLF_COLLISION_RADIUS: number = 74.0
    private CIRCLE_COLLISION_RADIUS: number = 78.0

    private static IsBeaconOfUnitedLifeforce = (r: Relic): r is BeaconOfUnitedLifeforce => {
        return r instanceof BeaconOfUnitedLifeforce
    }

    private static WolfCollisionFilter(k: Kitty): Func<bool> {
        return () => {
            return getFilterUnit().typeId == Constants.UNIT_CUSTOM_DOG && k.isAlive() && getFilterUnit().isAlive() // wolf should be alive too (exploding / stan wolf)
        }
    }

    private static ShadowRelicWolvesFilter(sk: ShadowKitty): Func<bool> {
        return () => {
            return GetUnitTypeId(getFilterUnit()) == Constants.UNIT_CUSTOM_DOG && sk.Unit.isAlive()
        }
    }

    private static ShadowRelicCircleFilter(sk: ShadowKitty): Func<bool> {
        return () => {
            return (
                GetUnitTypeId(getFilterUnit()) == Constants.UNIT_KITTY_CIRCLE &&
                getFilterUnit().owner != sk.Player && // Not Same Player
                sk.Unit.isAlive() && // Has to Be Alive
                Globals.ALL_KITTIES.get(getFilterUnit()!.owner).TeamID == Globals.ALL_KITTIES.get(sk.Player)!.TeamID
            ) // Must be same team
        }
    }

    private static CircleCollisionFilter(k: Kitty): Func<bool> {
        return () => {
            return (
                getFilterUnit().typeId == Constants.UNIT_KITTY_CIRCLE &&
                getFilterUnit().owner != k.Player && // Not Same Player
                k.isAlive() && // Has to Be Alive
                Globals.ALL_KITTIES.get(getFilterUnit()!.owner).TeamID == Globals.ALL_KITTIES.get(k.Player)!.TeamID && // Must be same team
                Gamemode.CurrentGameMode != GameMode.SoloTournament
            ) // Not Solo Mode
        }
    }

    public static KittyRegisterCollisions(k: Kitty) {
        let WOLF_COLL_RADIUS = k.CurrentStats.CollisonRadius
        k.w_Collision ??= Trigger.create()!
        k.c_Collision ??= Trigger.create()!

        UnitWithinRange.RegisterUnitWithinRangeTrigger(
            k.Unit,
            WOLF_COLL_RADIUS,
            WolfCollisionFilter(k),
            WolfCollisionTrigger(k)
        )
        UnitWithinRange.RegisterUnitWithinRangeTrigger(
            k.Unit,
            CIRCLE_COLLISION_RADIUS,
            CircleCollisionFilter(k),
            CircleCollisionTrigger(k)
        )
    }

    public static ShadowKittyRegisterCollision(sk: ShadowKitty) {
        sk.wCollision ??= Trigger.create()!
        sk.cCollision ??= Trigger.create()!

        UnitWithinRange.RegisterUnitWithinRangeTrigger(
            sk.Unit,
            DEFAULT_WOLF_COLLISION_RADIUS,
            ShadowRelicWolvesFilter(sk),
            WolfCollisionShadowTrigger(sk)
        )
        UnitWithinRange.RegisterUnitWithinRangeTrigger(
            sk.Unit,
            CIRCLE_COLLISION_RADIUS,
            ShadowRelicCircleFilter(sk),
            CircleCollisionShadowTrigger(sk)
        )
    }

    private static WolfCollisionTrigger(k: Kitty): Trigger {
        TriggerAddAction(k.w_Collision, () => {
            try {
                if (!k.Unit.isAlive()) return
                if (NamedWolves.ExplodingWolfCollision(getFilterUnit(), k)) return
                if (Globals.ALL_WOLVES[getFilterUnit()].IsReviving) return // bomber wolf
                if (ChronoSphere.RewindDeath(k)) return
                if (k.Invulnerable) return
                OneOfNine.OneOfNineEffect(k)
                k.KillKitty()
                TeamsUtil.CheckTeamDead(k)
            } catch (e: any) {
                Logger.Warning('Error: WolfCollisionTrigger: {e.Message}')
                throw e
            }
        })
        return k.w_Collision
    }

    private static CircleCollisionTrigger(k: Kitty): Trigger {
        TriggerAddAction(k.c_Collision, () => {
            try {
                let circle = Globals.ALL_KITTIES.get(getFilterUnit()!.owner)
                circle
                    .ReviveKitty(k)(k.Relics.find(CollisionDetection.IsBeaconOfUnitedLifeforce))
                    ?.BeaconOfUnitedLifeforceEffect(k.Player)
            } catch (e: any) {
                Logger.Warning('Error: CircleCollisionTrigger: {e.Message}')
                throw e
            }
        })
        return k.c_Collision
    }

    private static WolfCollisionShadowTrigger(sk: ShadowKitty): Trigger {
        TriggerAddAction(sk.wCollision, () => {
            if (NamedWolves.ExplodingWolfCollision(getFilterUnit(), sk.Kitty, true)) return // Floating text will appear on kitty instead of SK tho.
            sk.KillShadowKitty()
        })
        return sk.wCollision
    }

    private static CircleCollisionShadowTrigger(sk: ShadowKitty): Trigger {
        TriggerAddAction(sk.cCollision, () => {
            try {
                let circle = Globals.ALL_KITTIES.get(GetOwningPlayer(getFilterUnit()!))
                let saviorKitty = Globals.ALL_KITTIES.get(sk.Player)!
                circle.ReviveKitty(saviorKitty)
            } catch (e: any) {
                Logger.Warning('Error: CircleCollisionShadowTrigger: {e.Message}')
                throw e
            }
        })
        return sk.cCollision
    }
}
